src/Microsoft.ML.Data/DataLoadSave/TransformerChain.cs

```diff
@@ -146,7 +146,7 @@ public TransformerChain<TNewLast> Append<TNewLast>(TNewLast transformer, Transfo
             where TNewLast : class, ITransformer
         {
             Contracts.CheckValue(transformer, nameof(transformer));
-            return new TransformerChain<TNewLast>(_transformers.Append(transformer).ToArray(), _scopes.Append(scope).ToArray());
+            return new TransformerChain<TNewLast>(_transformers.AppendElement(transformer), _scopes.AppendElement(scope));
```

Comment:
Looks like this type only has enumerable constructor, so you end up with 3 arrays for each param.  One that is size N, and two that are size N+1.  If you keep it as an enumerable, you can reduce this.  It can be further reduced if you have a constructor that takes an array with reference semantics.

It's also somewhat odd that TransformerChain has copy semantics for constructor params, but EstimatorChain has reference semantics (and thus doesn't have this array copy problem).